htlcswitch/circuit.go

```diff
@@ -1,184 +1,307 @@
+type PaymentCircuit struct {
+	// AddRef is the forward reference of the Add update in the incoming
+	// link's forwarding package. This value is set on the htlcPacket of the
+	// returned settle/fail so that it can be removed from disk.
+	AddRef channeldb.AddRef
+
+	// Incoming is the circuit key identifying the incoming channel and htlc
+	// index from which this ADD originates.
+	Incoming CircuitKey
+
+	// Outgoing is the circuit key identifying the outgoing channel, and the
+	// HTLC index that was used to forward the ADD. It will be nil if this
+	// circuit's keystone has not been set.
+	Outgoing *CircuitKey
+
+	// PaymentHash used as unique identifier of payment.
+	PaymentHash [32]byte
+
+	// IncomingAmount is the value of the HTLC from the incoming link.
+	IncomingAmount lnwire.MilliSatoshi
+
+	// OutgoingAmount specifies the value of the HTLC leaving the switch,
+	// either as a payment or forwarded amount.
+	OutgoingAmount lnwire.MilliSatoshi
 
 	// ErrorEncrypter is used to re-encrypt the onion failure before
 	// sending it back to the originator of the payment.
 	ErrorEncrypter ErrorEncrypter
-}
 
-// circuitKey is a channel ID, HTLC ID tuple used as an identifying key for a
-// payment circuit. The circuit map is keyed with the identifier for the
-// outgoing HTLC
-type circuitKey struct {
-	chanID lnwire.ShortChannelID
-	htlcID uint64
+	// LoadedFromDisk is set true for any circuits loaded after the circuit
+	// map is reloaded from disk.
+	//
+	// NOTE: This value is determined implicitly during a restart. It is not
+	// persisted, and should never be set outside the circuit map.
+	LoadedFromDisk bool
 }
 
-// String returns a string representation of the circuitKey.
-func (k *circuitKey) String() string {
-	return fmt.Sprintf("(Chan ID=%s, HTLC ID=%d)", k.chanID, k.htlcID)
+// HasKeystone returns true if an outgoing link has assigned an this circuit's
+// outgoing circuit key.
+func (c *PaymentCircuit) HasKeystone() bool {
+	return c.Outgoing != nil
 }
 
-// CircuitMap is a data structure that implements thread safe storage of
-// circuit routing information. The switch consults a circuit map to determine
-// where to forward HTLC update messages. Each circuit is stored with its
-// outgoing HTLC as the primary key because, each offered HTLC has at most one
-// received HTLC, but there may be multiple offered or received HTLCs with the
-// same payment hash. Circuits are also indexed to provide fast lookups by
-// payment hash.
-//
-// TODO(andrew.shvv) make it persistent
-type CircuitMap struct {
-	mtx       sync.RWMutex
-	circuits  map[circuitKey]*PaymentCircuit
-	hashIndex map[[32]byte]map[PaymentCircuit]struct{}
+// newPaymentCircuit initalizes apayment circuit on the heap using the payment
+// hash and an in-memory htlc packet.
+func newPaymentCircuit(hash *[32]byte, pkt *htlcPacket) *PaymentCircuit {
+	var addRef channeldb.AddRef
+	if pkt.sourceRef != nil {
+		addRef = *pkt.sourceRef
+	}
+
+	return &PaymentCircuit{
```

Comment:
From my understanding:
```golang
// Allocates the circuit on the caller's stack.
// Compiler has option to optimize this to essentially be a macro
circuit1 := makePaymentCircuit()

// Invalid in golang, can't take reference of return value
invalidCircuitPtr := &makePaymentCircuit()

// Valid, but circuit is still allocated on caller's stack
circuitPtr1 := &circuit1

// Circuit is explictly allocated on heap
circuitPtr2 := newPaymentCircuit()
```
Caveat to `circuit1` is that if `circuit1Ptr` "escapes" the scope of the function, the compiler will allocate on the heap regardless. Howver, I have found this pattern to offer a little more control in how/where objects are allocated and make the intention clearer for the reader.

Primary benefit is that if you only need the `circuit1` within the current scope, you save an extra allocation by passing in `&circuit1`.

Reply:
Interesting, didn't know this! Did you benchmark this in any way?

Reply:
I haven't on this case specifically, combination of reading into golang's allocation/compiler optimizations and observations informed by benchmarks over the years :)

Reply:
FWIW, this pattern is particularly useful if you ever need to assign/append to slice of non-pointer types. Instead of doing
```golang
var things = make([]Thing, 8)
for i := range things {
    // Requires extra allocation and dereference
    things[i] = *NewThing()
}
```
the allocations can be sidestepped by
```golang
var things = make([]Thing, 8)
for i := range things {
    things[i] = MakeThing()
}
```
Having both types allows you to flexibly choose whether `[]*Thing` or `[]Thing` is more appropriate for the situation at hand